initial implementation of `grid`

t-bltg opened the pull request:
Naive implementation, after mention in https://github.com/FedeClaudi/Term.jl/issues/99#issuecomment-1140306519.

It looks like the tests on `master` or `v4` are completely broken so the workflow is not ideal to add tests covering this PR :/

This example works as expected:

```julia
using UnicodePlots, Term

panel(p; kw...) = Panel(p isa String ? p : string(p, color=true); fit=true, style="hidden", kw...)

main() = begin
  plots = collect(
    panel(lineplot(1:2, width=20, height=10, color=c))
    for c in (:red, :green, :blue, :yellow, :magenta, :cyan, :black, :white)
  )

  kw = (; placeholder=panel(""))

  # vector: auto layout
  for i in 2:length(plots)
    grid(plots[1:i]; kw...) |> display
  end

  # vector: explicit layout
  grid(plots, layout=(2, 4); kw...) |> display

  # matrix
  grid(reshape(plots[1:4], 2, 2)) |> display

  return
end

main()
```

- [ ] nest grids

FedeClaudi commented:
> It looks like the tests on master or v4 

Indeed, I had fixed the problem from master on `v4`, but then went ahead and made several changes that broke the tests and haven't updated the tests yet. I can try to do this in the next couple days which might make development easier here.

A few considerations:
1. Perhaps there could be a way to let the user specify the `aspect_ratio`  
2. I see users can pass `layout` specifying the number of columns and rows. It might be nice to let the user specify one of them and compute the other based on the aspect ratio and number of renderables. 
3. Add padding between renderables. Use `Spacer` or a string to add vertical/horizontal spacing between renderables (with a user specified amount)
4 I wander if it would be worth making a `Grid` mutable type that stores information about the grid's layout and it's content. This way you can first create a grid and then fill it up with renderables as you go. 
5 for prototyping (and debugging for us) it might be nice to be able to create a grid without having to provide renderables (e.g. with just place holders)

Finally, about the nested grids: if the object returned by `grid` is an abstract renderable that's just another renderable that can go in a second grid.

t-bltg commented:
Thanks for the comments and the dev hints !

> If any of this goes beyond what you wanted to get out of grid for your own needs

Nah, I'm fine with trying to implement these, but I'm very much relying on tests for development (they go with the dev flow, increasing complexity only after covering basic tests), so I'll come back to this PR when running the tests using `] test Term` will be fixed. Is that OK for you ?

FedeClaudi commented:
Yeah of course, I'll start working on those asap!

FedeClaudi commented:
I couldn't help myself and wanted to try implementing something before looking at the tests...

As I said above it might be nice to be able to create an "empty" grid (for prototyping stuff or to fill it in later) - and for that I thought It'd be nice to have a nice `place holder` renderable. So I made one:
<img width="303" alt="image" src="https://user-images.githubusercontent.com/17436313/170864866-efcd6ce9-272a-43c9-8c43-1634222a60b2.png">

<img width="559" alt="image" src="https://user-images.githubusercontent.com/17436313/170864918-409bef6f-3b5e-4a32-a8ff-bbaa94959d83.png">

Also note how I've used empty strings for padding. The `hstack, vstack` code is smart enough to correctly stack the renderables with the string as spacing!

It's on `v4` if you want to look at it - now I get to the tests I promise 😬

t-bltg commented:
> I couldn't help myself and wanted to try implementing something before looking at the tests...

Haha great, for your defense I do the same when working on `UnicodePlots`. You're definitively the expert here, and I don't know much on the internals of `Term` so you definitively have a clearer view on this ... I cannot work on this today, but I will for sure in the next couple of days ;)

FedeClaudi commented:
All tests passing locally on `v4` - I'm running them on https://github.com/FedeClaudi/Term.jl/pull/102

t-bltg commented:
I've updated the branch (btw as a maintainer, you can push in this branch too even if it is not attached to your repo).

I couldn't help noticing some inconsistencies with the code style used in `Term.jl`, just my 2 cents:

1. modules should start with a capital letter and use `CamelCase` https://docs.julialang.org/en/v1/manual/style-guide/#Use-naming-conventions-consistent-with-Julia-base/
2. using `eval` can be replaced with `getfield(Module, Symbol)`

FedeClaudi commented:
Hey, Nice to see the development progress so rapidly here!

Thanks for the help and advice, I'm still relatively new to Julia and it shows...

Regarding your points above:
1. Yeah.. I think I've been resisting it because of my Python background - but I should conform to standard practice. I'm introducing some breaking changes with the next release (`v4`), so now is probably a good time to do this. 
2. I didn't know about that usage! I will update that
3. I have the formatter running as part of the [`CI`](https://github.com/FedeClaudi/Term.jl/blob/032f6a5200b84aa17d0b1c48d6453502729872c7/.github/workflows/CI.yml#L54) github action, and have the [toml file](https://github.com/FedeClaudi/Term.jl/blob/v4/.JuliaFormatter.toml), but this could be expanded with more arguments like yours.  But in theory the formatter should already be running for PRs (see [here](https://github.com/FedeClaudi/Term.jl/runs/6646479991?check_suite_focus=true)?

FedeClaudi commented:
Thanks for adding the test for `PlaceHolder`. I'm finishing up the new `Table` type and then I will be adding a ton more tests since the coverage dropped a bit recently.

FedeClaudi commented on src/layout.jl:
```diff
@@ -635,36 +630,67 @@ PlaceHolder(ren::AbstractRenderable) = PlaceHolder(ren.measure.w, ren.measure.h)
     num = length(rens)
-    if layout === nothing
-        m = ceil(Int, √(num))
+    m, n = layout === nothing ? calc_nrows_ncols(num, aspect) : layout
```
Trying to learn something here - your Julia is definitely much better than mine!
What's the advantage of using `layout === nothing` over `isnothing(layout)`?

t-bltg replied:
AFAIK , in the old days, `=== nothing` was better for the compiler. Also, using this syntax was to retain compatibility with the older julia versions, before `isnothing` was introduced. So nowadays, it is only a matter of personal taste and code style: I see now that your are already using `isnothing`, I will change to that.

FedeClaudi replied:
I see, thanks!

johnnychen94 replied:
> What's the advantage of using layout === nothing over isnothing(layout)?

They're exactly the same thing if you check `@less isnothing(1)`. One small difference is that `isnothing` is introduced in Julia 1.1 so if you're going to maintain 1.0 compatibility, you should use `===`. But since Term.jl requires Julia at least 1.6, just using `isnothing` should be fine -- people introduce and recommend `isnothing` because it's not immediately obvious to new Julia users that `x === nothing` is faster than `x == nothing`.

In Julia >= 1.7, there's also an equivalent version `m, n = @something layout calc_nrows_ncols(num, aspect)` that returns the first non-nothing value.

FedeClaudi commented on src/layout.jl:
```diff
@@ -635,36 +630,67 @@ PlaceHolder(ren::AbstractRenderable) = PlaceHolder(ren.measure.w, ren.measure.h)
+Computes the number of rows and columns to fit a given number `n` of subplots in a figure with aspect `aspect`.
+Adapted from: stackoverflow.com/a/43366784
+"""
+function calc_nrows_ncols(n, aspect=(16, 9))
```
Might be nice to also allow for the aspect ratio to be a single number (e.g. `1` for a square grid)

t-bltg replied:
Done.

FedeClaudi replied:
thanks!

johnnychen94 commented on src/layout.jl:
```diff
@@ -635,36 +630,67 @@ PlaceHolder(ren::AbstractRenderable) = PlaceHolder(ren.measure.w, ren.measure.h)
-function grid(rens::AbstractMatrix{<:AbstractRenderable})
-    rows = collect(foldl(*, col[2:end], init = first(col)) for col in eachrow(rens))
-    return foldl(/, rows[2:end], init = first(rows))
+function grid(rens::AbstractMatrix{<:AbstractRenderable}; hspacer = "", vspacer = hspacer)
```
Unlike Python, the default values in Julia are set when called instead of when defined:

```julia
julia> f(x, y=x) = y
f (generic function with 2 methods)

julia> f(1)
1

julia> f(2)
2
```

FedeClaudi replied:
Sorry if I can suggest one additional change, I think vertical/horizontal spacing (or padding) should be expressed as integers: the amount of space required. It would make more sense for the user, then you can create the strings under the hood.

FedeClaudi replied:
Hey @johnnychen94  nice to see you here! I understand now, beautiful - thanks

t-bltg replied:
> you can add a previous keyword argument as the value for another kehword argument

It's a common thing in julia to apply a function on given keyword to compute a default value e.g. https://github.com/JuliaPlots/Plots.jl/blob/b959e64b9c189c1e5383582f7d1963a27e76c975/src/axes.jl#L594.

> Can they be set independently

yes.

> instead of h/vspacer

Renamed to `hpad`, `vpad`.

> I think vertical/horizontal spacing (or padding) should be expressed as integers:

Will fix.

FedeClaudi replied:
I had no idea! I can see how that could be very useful, nice to know!

And thanks for making the changes

t-bltg commented:
> My only problem is that I e.g. have types like Box and Panel that I want to have in a module by the same name. Is it allowed to have a type Panel in a module Panel? That would probably cause confusion too?

One possibility (certainly not the only one) would be to add a `Type` suffix (e.g. https://github.com/JuliaPlots/UnicodePlots.jl/blob/bf6f643d6d99036a08e2fc5229943ba9b75e3381/src/common.jl#L167) ?

> I have the formatter running as part of the [CI](https://github.com/FedeClaudi/Term.jl/blob/032f6a5200b84aa17d0b1c48d6453502729872c7/.github/workflows/CI.yml#L54) github action, and have the [toml file](https://github.com/FedeClaudi/Term.jl/blob/v4/.JuliaFormatter.toml)

Huh, I'm blind I need another pair of glasses :see_no_evil:

FedeClaudi commented on src/layout.jl:
```diff
@@ -635,36 +630,67 @@ PlaceHolder(ren::AbstractRenderable) = PlaceHolder(ren.measure.w, ren.measure.h)
-    return foldl(/, rows[2:end], init = first(rows))
+function grid(rens::AbstractMatrix{<:AbstractRenderable}; hspacer = "", vspacer = hspacer)
+    rows = collect(foldl((a, b) -> a * hspacer * b, col[2:end], init = first(col)) for col in eachrow(rens))
+    return foldl((a, b) -> a * vspacer * b, rows[2:end], init = first(rows))
```
If I understand correctly here is where you're stacking single rows of the grid, shouldn't it be `/` instead of `*`.

If you have multiple rows, each of which is a `Renderable` you can use `vstack(rows...)` to vertically stack all of them at once, but I guess you;d have to manually insert the `vspacer` in there. 


One alternative option whould be to create the columns first and then horizontally stack them to get the grid. The advantage with this is that you could use `cvstack` to ensure renderables in each column are centered. It might help with the layout for renderables of unequal sizes?

t-bltg replied:
> If I understand correctly here is where you're stacking single rows of the grid, shouldn't it be / instead of *.

Copy-paste mistake, fixed.

FedeClaudi commented:
> One possibility (certainly not the only one) would be to add a Type suffix (e.g. https://github.com/JuliaPlots/UnicodePlots.jl/blob/bf6f643d6d99036a08e2fc5229943ba9b75e3381/src/common.jl#L167) ?

If I understand correctly that's for a `Type`, I meant the `struct`s that make up the different renderables. Like here you have a [`Plot`](https://github.com/JuliaPlots/UnicodePlots.jl/blob/bf6f643d6d99036a08e2fc5229943ba9b75e3381/src/plot.jl#L47) struct, if you had to put that in a module also called `plot` it would be tricky no?

FedeClaudi commented:
Sorry for the many comments, I feel like maybe I'm micromanaging this PR too much! 
I really appreciate your contribution to your project, feel free to ignore and submit what you think is appropriate - I can make any small change later on if necessary.

t-bltg commented:
> But in theory the formatter should already be running for PRs (see [here](https://github.com/FedeClaudi/Term.jl/runs/6646479991?check_suite_focus=true)?

Ha ok, i see now the format only warns, so it is useless (no one read the logs), I think it should error the `format` job ?

FedeClaudi commented:
Oh, I thought it worked like `pre-commit + black` in python where it actually formats the code - is it just checking that the formatting is ok? Is there a way to do automatic formatting?

t-bltg commented:
>  Is there a way to do automatic formatting

There is but IMO, the best option is to force people to run julia formatter on their PRs. The issue with auto-format PRs is that they pollute the git history of your repo (as @johnnychen94 mentionned here: https://github.com/JuliaPlots/UnicodePlots.jl/pull/207#pullrequestreview-858005720).

FedeClaudi commented:
I see. I'd love to have something like `pre-commit` hooks running the formatter (or doing other checks like [flake8](https://gitlab.com/pycqa/flake8) in python). Is there an option for that? I'm happy to implement whatever is considered best practice here.

FedeClaudi commented:
>  being consistent with a clear code style is imho better for end users or dev newcomers ?

I completely agree! I just wish the tooling was a bit better - in python with `pre-commit` + hooks you can't even push a commit until you've fixed stuff. You can enforce formatting + other standard (e.g. avoid importing stuff that is not used) which is great. It also means you don't have to remember to run the formatter manually.. 

I thought the `format` option in the CI was actually doing the formatting, not just checking. I will start running it manually every now and then too!

t-bltg commented:
They talk about `pre-commit` here https://discourse.julialang.org/t/ann-juliaformatter-jl/48084/15 but there doesn't seem to be any implemented.

FedeClaudi commented:
Sounds good. 

Well, I've learned  a lot from this PR!

I like where the development of `grid` is going, let me know when you feel like it's ready for merging. 
I'd like to merge this in `v4` (although I might have a headache due to the formatting changes) and then merge the whole thing in `master` before the next release (in 1-2 weeks I think). 

Feel free to add tests for `grid` if you'd like, but I can also do that later on.

FedeClaudi commented:
> a dev is always lazy

Indeed we are! Working in pairs helps with that though (at least for me). 

> I'm going to try the nesting of grids now.

This should be easy I think: if the return grid is a `Renderable` and inputs to the grid are `Renderables`, then you can just pass a grid as an argument to another `grid` call. 

> if the return grid is a `Renderable` 

It should be - because of the calls to `v/hstack` in the grid making function. This also means that you can use the grid with normal layout commands (e.g. have a grid in a panel - hstack a grid with another renderable etc...).

t-bltg commented:
Quick question: are you ok with changing definitions like these https://github.com/FedeClaudi/Term.jl/blob/fe8525ffcbc9f8f38b56465ed415a844795f2480/src/segment.jl#L30 to single line: `Segment(text) = Segment(apply_style(text), Measure(text))` ?

t-bltg commented:
> My only problem is that I e.g. have types like Box and Panel that I want to have in a module by the same name. Is it allowed to have a type Panel in a module Panel? That would probably cause confusion too?

So e.g. rename module `segment` to `Segment` fails because it exports `Segment`. Should we name the module `Segments` (plural) instead ?

FedeClaudi commented:
I'll put some thought here which I think @johnnychen94  might be interested in - but I don't think they should be implemented in this PR. 

Currently `grid` like all other "layout" methods returns a `Renderable` with its own `segments` and `measure`. This means that any information of what went into the layout is lost: you can't change an element within the layout. The layout is a single object. 

Thus, it might be worth having a non-destructive layout approach: create a layout object that stores a bunch of renderables and their relationship (panel A goes in the first grid element, panel B in the second...). Then, when you want to print this layout object you create the appropriate renderable. But this would give you the option to e.g. change panel A (`layout[1] = newpanel`) before re-printing the whole thing. 
I'm thinking this would be especially useful for interactive things where you want to update the output to the user and you want to be as efficient as possible. 

One question would be @johnnychen94  if this is something that should be done as part of the GSOC `TermLayout` project or in Term? Happy to hear your thoughts here...

FedeClaudi commented:
> Quick question: are you ok with changing definitions like these 

Yes, always one-liners when possible. I think that was changed by the formatting? We can disable that in the `.toml` file right? I like the options you have for `UnicodePlots.jl`

? So renaming for example module segment to Segment fails because it exports Segment. Should we name the module Segments (plural) instead ?

It wouldn't be my first choice, but I guess consistency trumps my personal preference here. We can make all modules `CamelCase` and plural (`Tables, Panels, Boxes`)...

t-bltg commented:
> I think that was changed by the formatting?

It seems not, I will do that manually. There seems to be a [`short_to_long_function_def`](https://github.com/domluna/JuliaFormatter.jl/blob/master/README.md#short_to_long_function_def) but no reverse (that would be very hard to implement ?) ...

FedeClaudi commented:
I see, thanks for taking care of all that. You can also submit the PR as is and I can handle these boring tasks

johnnychen94 commented:
> One question would be @johnnychen94 if this is something that should be done as part of the GSOC TermLayout project or in Term? Happy to hear your thoughts here...

I think it's okay. It doesn't matter much where trivial features like this are added, and there's no need to block this nice feature only because we have a GSoC student planning the TermLayouts stuff.

FedeClaudi commented:
Got it. I've added it to the to-do list for this release then! 
I guess if it works nicely it's something that the student can make use of and he can focus on the more challenging stuff like the REPL business - but that's a different topic we can discuss elsewhere, thanks for the feedback!

FedeClaudi commented:
Hey, @t-bltg 

As I mentioned, I've been implementing a `Table` renderable in `v4`.
It just dawned on me that this can handle renderables too when creating a table:
<img width="411" alt="image" src="https://user-images.githubusercontent.com/17436313/170988740-f58f8025-7b08-4a19-a16d-ae90822fe85e.png">

Now, `Table` can already take care of creating a layout of renderables handling horizontal (and soon vertical) padding, left/center/right column-wise justification ...
<img width="522" alt="image" src="https://user-images.githubusercontent.com/17436313/170989016-2868df07-256d-46d1-b529-f8c55c4838ca.png">


... pretty much the same stuff that `grid` need to take care of as well. 
So maybe instead of duplicating functionality, `grid` could create a `Table` renderable by making an appropriate call to `Table` instead of doing the layout directly. As in, compute the number of columns/rows and organize the renderables into a matrix that is then passed to `Table`. Then `Table` takes care of padding/justification etc... 
The table edges could be hidden using your new `:HIDDEN` `Box` instance.  

What do you think?

t-bltg commented:
I've managed to properly rename all modules, using plural only if necessary, that took some time.

Some tiny changes:
`Tables`, use `import Tables as TablesPkg`
`Logging` renamed to `Logs`

I've fixed all the docs (broken links, refs, ...) which took a long time as well.

So what I propose is that we merge this now, so that you can work on `v4` without dealing with / creating tons of merge conflicts:
I've removed `grid` for now, and I can start working on it again in another PR, after https://github.com/FedeClaudi/Term.jl/pull/101#issuecomment-1141077468 ...

t-bltg commented:
> I can give you a head's up here when that's ready

OK, since the CI is green on this PR, I'll stop working on it for now and wait for the head's up.

FedeClaudi commented:
Hey @t-bltg  The code at https://github.com/FedeClaudi/Term.jl/pull/102 has the finished up version of `Table`.

I've added a script in `examples` showcasing how to use it a bit. 
But essentially for `grid` you'd need to create a `Matrix` of renderables and then call `Table` with it. If you set `show_header` as False and `box=:HIDDEN` the result should be the grid. 
You can also use `vpad` and `hpad` as well as `columns_justify` and `vertical_justify`. 

Here's a small example

```julia
using Term.Tables
import Term.Layout: PlaceHolder

# t = 1:3
# data = hcat(t, rand(Int8, length(t)))

p = PlaceHolder(25, 10)

data = [
    [p p p]
    [p p p]
]

Table(data; box=:NONE, header=["ONE", "TWO", "THREE"],  vpad=0, hpad=2, header_style="bold green")
```

<img width="539" alt="image" src="https://user-images.githubusercontent.com/17436313/171049468-210b5d3d-6588-473a-9e7d-249b81d12d5a.png">

t-bltg commented:
Great ! I'm currently afk, so I'll have a look tomorrow evening.
